src/Compiler/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorTagHelperContextDiscoveryPhase.cs

```diff
@@ -263,6 +266,8 @@ public ComponentDirectiveVisitor(string filePath, IReadOnlyList<TagHelperDescrip
                     Matches.Add(tagHelper);
                 }
             }
+
+            _notFullyQualifiedComponents = builder.ToImmutable();
```

Comment:
This is a guaranteed array copy. We should consider adding an extension `DrainToImmutable` on our `ImmutableArrayExtensions` that mimics the one the framework is shipping in .NET 8: https://source.dot.net/#System.Collections.Immutable/System/Collections/Immutable/ImmutableArray_1.Builder.cs,216.

Reply:
(Though perhaps with just clearing the array, trying to resize, if it's not the right size and must be copied anyway)

Reply:
@davidwengier: If you decide to add that extension method, there's a version [here](https://github.com/dotnet/razor/blob/e7f48ca663e3881d775e52fa37f211064f23e312/src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/PooledObjects/PooledArrayBuilder%601.cs#L71-L106) that you can use as a guide.

Reply:
May as well just use that then. Gives me an opportunity to complain that we have an ArrayBuilderPool and a PooledArrayBuilder :P